trunk/SUAVE/Components/Energy/Storages/Batteries/Constant_Mass/Lithium_Ion_LiFePO4_18650.py

```diff
@@ -0,0 +1,362 @@
+        E_bat = np.dot(I,P)
+        E_bat = np.reshape(E_bat,np.shape(E_current)) #make sure it's consistent
+        
+        # Add this to the current state
+        if np.isnan(E_bat).any():
+            E_bat=np.ones_like(E_bat)*np.max(E_bat)
+            if np.isnan(E_bat.any()): #all nans; handle this instance
+                E_bat=np.zeros_like(E_bat)
+                
+        E_current = E_bat + E_current[0]
+        
+        SOC_new = np.divide(E_current,E_max)
+        SOC_new[SOC_new>1] = 1.
+        SOC_new[SOC_new<0] = 0. 
+        DOD_new = 1 - SOC_new
+          
+        # Determine new charge throughput (the amount of charge gone through the battery)
+        Q_total    = np.atleast_2d(np.hstack(( Q_prior[0] , Q_prior[0] + cumtrapz(abs(I_bat)[:,0], x   = numerics.time.control_points[:,0])/Units.hr ))).T      
+                
+        # A voltage model from Chen, M. and Rincon-Mora, G. A., "Accurate Electrical Battery Model Capable of Predicting
+        # Runtime and I - V Performance" IEEE Transactions on Energy Conversion, Vol. 21, No. 2, June 2006, pp. 504-511
+        V_normalized  = (-1.031*np.exp(-35.*SOC_new) + 3.685 + 0.2156*SOC_new - 0.1178*(SOC_new**2.) + 0.3201*(SOC_new**3.))/4.1
+        V_oc = V_normalized * V_max
+        V_oc[ V_oc > V_max] = V_max
+        
+        # Voltage under load:
+        if battery_discharge_flag:
+            V_ul   = V_oc - I_bat*R_0
+        else: 
+            V_ul   = V_oc + I_bat*R_0 
+             
+        # Pack outputs
+        battery.current_energy                     = E_current
+        battery.resistive_losses                   = Q_heat_gen
+        battery.cell_temperature                   = T_current 
+        battery.pack_temperature                   = T_current 
+        battery.load_power                         = V_ul*I_bat
+        battery.state_of_charge                    = SOC_new 
+        battery.depth_of_discharge                 = DOD_new
+        battery.charge_throughput                  = Q_total 
+        battery.cell_charge_throughput             = Q_total/n_parallel  
+        battery.voltage_open_circuit               = V_oc
+        battery.voltage_under_load                 = V_ul
+        battery.current                            = I_bat 
+        battery.pack_temperature                   = T_current 
+        battery.cell_joule_heat_fraction           = np.zeros_like(V_ul)
+        battery.cell_entropy_heat_fraction         = np.zeros_like(V_ul)  
+        battery.cell_voltage_open_circuit          = np.zeros_like(V_ul)
+        battery.cell_current                       = np.zeros_like(V_ul)
+        battery.thevenin_voltage                   = np.zeros_like(V_ul)
+        battery.heat_energy_generated              = Q_heat_gen 
+        battery.internal_resistance                = R_0 
+        battery.cell_voltage_under_load            = V_ul 
+        
+        return      
+    
+    def append_battery_unknowns(self,segment): 
```

Comment:
same comment; function already exists in base class

Reply:
Addressed